File: src/window.zig
//! Window query and control functions. Wraps `sokol.app.*` so user code never
//! needs to import sokol directly.
const sapp = @import("sokol").app;
const pxl = @import("pxl.zig");

pub fn width() i32 {
    return sapp.width();
}
pub fn height() i32 {
    return sapp.height();
}
pub fn widthf() f32 {
    return sapp.widthf();
}
pub fn heightf() f32 {
    return sapp.heightf();
}
pub fn dpiScale() f32 {
    return sapp.dpiScale();
}
pub fn isFullscreen() bool {
    return sapp.isFullscreen();
}
pub fn toggleFullscreen() void {
    sapp.toggleFullscreen();
}
pub fn showMouse(show: bool) void {
    sapp.showMouse(show);
}
pub fn mouseShown() bool {
    return sapp.mouseShown();
}
pub fn lockMouse(lock: bool) void {
    sapp.lockMouse(lock);
}
pub fn mouseLocked() bool {
    return sapp.mouseLocked();
}
pub fn requestQuit() void {
    sapp.requestQuit();
}
pub fn cancelQuit() void {
    sapp.cancelQuit();
}
pub fn quit() void {
    sapp.quit();
}
pub fn setWindowTitle(title: [:0]const u8) void {
    sapp.setWindowTitle(title);
}
pub fn setClipboardString(str: [:0]const u8) void {
    sapp.setClipboardString(str);
}
pub fn getClipboardString() [:0]const u8 {
    return sapp.getClipboardString();
}

// ── Render target ────────────────────────────────────────────────────────────

/// True when the resolution policy guarantees an integer-scaled blit of a fixed
/// design-size render target.
pub fn isPixelPerfect() bool {
    return switch (pxl.gpu.gfx_config.resolution_policy) {
        .no_border_pixel_perfect, .show_all_pixel_perfect => true,
        else => false,
    };
}

pub fn renderWidth() i32 {
    const scaler = pxl.gpu.gfx_config.resolution_policy.getScaler(
        pxl.gpu.gfx_config.design_width,
        pxl.gpu.gfx_config.design_height,
    );
    return scaler.w;
}

pub fn renderHeight() i32 {
    const scaler = pxl.gpu.gfx_config.resolution_policy.getScaler(
        pxl.gpu.gfx_config.design_width,
        pxl.gpu.gfx_config.design_height,
    );
    return scaler.h;
}

pub fn renderWidthf() f32 {
    return @floatFromInt(renderWidth());
}

pub fn renderHeightf() f32 {
    return @floatFromInt(renderHeight());
}

File: rust/pxl/src/window.rs
//! Window query and control functions — mirrors `pxl.window.*` in Zig.

use std::ffi::CStr;
use pxl_sys;

pub fn width() -> i32 {
    unsafe { pxl_sys::pxl_window_width() }
}
pub fn height() -> i32 {
    unsafe { pxl_sys::pxl_window_height() }
}
pub fn widthf() -> f32 {
    unsafe { pxl_sys::pxl_window_widthf() }
}
pub fn heightf() -> f32 {
    unsafe { pxl_sys::pxl_window_heightf() }
}
pub fn dpi_scale() -> f32 {
    unsafe { pxl_sys::pxl_window_dpi_scale() }
}
pub fn is_fullscreen() -> bool {
    unsafe { pxl_sys::pxl_window_is_fullscreen() }
}
pub fn toggle_fullscreen() {
    unsafe { pxl_sys::pxl_window_toggle_fullscreen() }
}
pub fn show_mouse(show: bool) {
    unsafe { pxl_sys::pxl_window_show_mouse(show) }
}
pub fn mouse_shown() -> bool {
    unsafe { pxl_sys::pxl_window_mouse_shown() }
}
pub fn lock_mouse(lock: bool) {
    unsafe { pxl_sys::pxl_window_lock_mouse(lock) }
}
pub fn mouse_locked() -> bool {
    unsafe { pxl_sys::pxl_window_mouse_locked() }
}
pub fn request_quit() {
    unsafe { pxl_sys::pxl_window_request_quit() }
}
pub fn cancel_quit() {
    unsafe { pxl_sys::pxl_window_cancel_quit() }
}
pub fn quit() {
    unsafe { pxl_sys::pxl_window_quit() }
}
pub fn set_title(title: &str) {
    let c_str = std::ffi::CString::new(title).unwrap();
    unsafe { pxl_sys::pxl_window_set_title(c_str.as_ptr()) }
}
pub fn set_clipboard(text: &str) {
    let c_str = std::ffi::CString::new(text).unwrap();
    unsafe { pxl_sys::pxl_window_set_clipboard(c_str.as_ptr()) }
}
pub fn get_clipboard() -> &'static str {
    unsafe {
        let ptr = pxl_sys::pxl_window_get_clipboard();
        if ptr.is_null() {
            return "";
        }
        CStr::from_ptr(ptr).to_str().unwrap_or("")
    }
}